tikv/raftstore/worker.go

```diff
@@ -285,27 +289,182 @@ type pendingDeleteRanges struct {
+	}
+}
+
+/// ingestMaybeStall checks the number of files at level 0 to avoid write stall after ingesting sst.
+/// Returns true if the ingestion causes write stall.
+func (snapCtx *snapContext) ingestMaybeStall() bool {
+	for _, cf := range snapshotCFs {
+		if plainFileUsed(cf) {
+			continue
+		}
+		// todo, related to cf.
+	}
+	return false
+}
+
+// cleanupOverlapRanges gets the overlapping ranges and cleans them up.
+func (snapCtx *snapContext) cleanUpOverlapRanges(startKey, endKey []byte) {
+	overlapRanges := snapCtx.pendingDeleteRanges.drainOverlapRanges(startKey, endKey)
+	useDeleteFiles := false
+	for _, r := range overlapRanges {
+		snapCtx.cleanUpRange(r.regionId, r.startKey, r.endKey, useDeleteFiles)
+	}
+}
+
+// insertPendingDeleteRange inserts a new pending range, and it will be cleaned up with some delay.
+func (snapCtx *snapContext) insertPendingDeleteRange(regionId uint64, startKey, endKey []byte) bool {
+	if int64(snapCtx.cleanStalePeerDelay.Seconds()) == 0 {
+		return false
+	}
+	snapCtx.cleanUpOverlapRanges(startKey, endKey)
+	log.Infof("register deleting data in range. [regionId: %d, startKey: %s, endKey: %s]", regionId,
+		regionId, hex.EncodeToString(startKey), hex.EncodeToString(endKey))
+	timeout := time.Now().Add(snapCtx.cleanStalePeerDelay)
+	snapCtx.pendingDeleteRanges.insert(regionId, startKey, endKey, timeout)
+	return true
+}
+
+// cleanUpRange cleans up the data within the range.
+func (snapCtx *snapContext) cleanUpRange(regionId uint64, startKey, endKey []byte, useDeleteFiles bool) {
+	// todo, currently it is a place holder.
+}
+
 type regionRunner struct {
 	ctx *snapContext
 	// we may delay some apply tasks if level 0 files to write stall threshold,
 	// pending_applies records all delayed apply task, and will check again later
 	pendingApplies []task
+	wg             sync.WaitGroup
 }
 
 func newRegionRunner(engines *Engines, mgr *SnapManager, batchSize uint64, cleanStalePeerDelay time.Duration) *regionRunner {
-	return nil // TODO: stub
+	return &regionRunner{
+		ctx: &snapContext{
+			engiens:             engines,
+			mgr:                 mgr,
+			batchSize:           batchSize,
+			cleanStalePeerDelay: cleanStalePeerDelay,
+			pendingDeleteRanges: &pendingDeleteRanges{},
+		},
+	}
+}
+
+// handlePendingApplies tries to apply pending tasks if there is some.
+func (r *regionRunner) handlePendingApplies() {
+	for len(r.pendingApplies) != 0 {
+		// Should not handle too many applies than the number of files that can be ingested.
+		// Check level 0 every time because we can not make sure how does the number of level 0 files change.
+		if r.ctx.ingestMaybeStall() {
+			break
+		}
+		task := r.pendingApplies[0].data.(*regionTask)
+		r.pendingApplies = r.pendingApplies[1:]
+		r.ctx.handleApply(task.regionId, task.status)
+	}
+}
+
+func (r *regionRunner) handleGen(regionTask *regionTask) {
+	go r.ctx.handleGen(regionTask.regionId, regionTask.raftSnapshot, regionTask.kvSnapshot, regionTask.notifier)
```

Comment:
For now, we can just do it without goroutine.

Reply:
ok